doc/authentication.md

```diff
@@ -0,0 +1,73 @@
+Requiring authentication
+========================
+
+In some cases, it might be needed for security or legal reasons to secure the
+projects' code available through CodeCompass against unauthorised access.
+
+To enable this feature, an `authentication.json` file should be created under
+the workspace directory (`--workspace` or `-w` flag given to the server).
+
+At a bare minimum, to restrict access, an `"enabled": true` MUST be present
+in the JSON.
+
+An example valid JSON looks like this:
+
+~~~~{.json}
+{
+    "enabled": true,
```

Comment:
authentication.enabled (the topmost one) toggles authentication as a whole on the server.
enabled for each backend turns the backend on and off. This is done so that backends can be toggled without having to remove their already existing configuration from the file.

If global auth is ON, but there are no working backends (they are off, couldn't load the binary, or there are config errors, whatsoever) the server will tell the user that there weren't any usable auth backends and that the administrator should fix the configuration.

The server does **NOT** fall back to unauthenticated mode to protect the data from access. Instead, there will be authentication required, but no matter what you try to authenticate with, you'll never be able to get in, until the server's config is fixed and the server is restarted.